### PR TITLE
Build Scripts for CI

### PR DESCRIPTION
**Pull request by csibbitt**
* Moved common vars to metadata.sh
  * Went a bit overboard
  * Resulting build/push/deploy scripts should be ~useful for other operators
* s/SAF_PROJECT/OCP_PROJECT/g a bunch of places for generic/consistency
* Automated local build (okay it was a one-liner)
* Automated local push & deploy via CSV
  * Thanks to @mrunge and @leifmadsen for tag-teaming until we figured this out
* Added relative pathing to quickstart
  * Now you can run it like ./deploy/quickstart.sh (or from anywhere)
* Removed install-and-run-ocp.sh, find it on ocp3 branch if you need it

### CODE REVIEW

**Comment by csibbitt** [comment]
(At least) two deficiencies:
* Still lacks an "It's okay to try the smoketest now" wait-state
* I'd like to do better than this and find a way to install a subscription that points to an alternate CSV, that way we can properly test the dependency resolution that OLM does as well as eliminate the weird re-install shuffle. If anyone figures out a working formula to do that, please LMK! I imagine it could be done by staging every build into quay or something but it seems very wasteful.


# Testing

## First try

* First I just ran it

```
[csibbitt@laptop service-assurance-operator (csibbitt-900-ci-scripts) ]$ ./build/build_ci.sh 
Building...
INFO[0000] Building OCI image service-assurance-operator:latest 
[..]
INFO[0005] Operator build complete.                     
Removing old SAF...
[...]
Waiting for sa-telemetry to disappear
Waiting for sa-telemetry to disappear
Error from server (NotFound): namespaces "sa-telemetry" not found
Running quickstart...
[...]
operatorsource.operators.coreos.com/redhat-service-assurance-operators created
subscription.operators.coreos.com/amq7-cert-manager created
operatorgroup.operators.coreos.com/sa-telemetry-og created
subscription.operators.coreos.com/serviceassurance-operator-alpha-redhat-service-assurance-operators-openshift-marketplace created
No resources found.
waiting for SAO...
[...]
waiting for SAO...
service-assurance-operator.v0.1.1   Service Assurance Operator   0.1.1                                 Succeeded
Pushing new operator image...
Login Succeeded!
[...]
Writing manifest to image destination
Storing signatures
Re-deploying with local build...
subscription.operators.coreos.com "serviceassurance-operator-alpha-redhat-service-assurance-operators-openshift-marketplace" deleted
clusterserviceversion.operators.coreos.com "service-assurance-operator.v0.1.1" deleted
serviceaccount/service-assurance-operator created
role.rbac.authorization.k8s.io/service-assurance-operator created
rolebinding.rbac.authorization.k8s.io/service-assurance-operator created
clusterserviceversion.operators.coreos.com/service-assurance-operator.v0.1.1 created
serviceassurance.infra.watch/saf-default created

[csibbitt@laptop service-assurance-operator (csibbitt-900-ci-scripts) ]$ echo $?
0

[csibbitt@laptop service-assurance-operator (csibbitt-900-ci-scripts) ]$ oc describe po service-assurance-operator-c8b7f986f-fz6kf | grep image                                                                
    Image:         image-registry.openshift-image-registry.svc:5000/sa-telemetry/service-assurance-operator:latest
    Image ID:      image-registry.openshift-image-registry.svc:5000/sa-telemetry/service-assurance-operator@sha256:040b54edd5c8bc3668960455e13a65880ce00edbfe9ff462a4e18b39325679da
```

* In the final output above, we can see that the SAO has been deployed with the image pushed into the OCP registry instead of the one from quay

* Now check on the rest of SAF

```
[csibbitt@laptop service-assurance-operator (csibbitt-900-ci-scripts) ]$ oc get po
NAME                                          READY   STATUS      RESTARTS   AGE
alertmanager-saf-default-0                    2/2     Running     0          3m37s
interconnect-operator-55b5f8867b-2lzfq        1/1     Running     0          5m9s
prometheus-operator-68f97b8f79-5cnzl          1/1     Running     0          5m17s
prometheus-saf-default-0                      3/3     Running     1          3m41s
saf-default-interconnect-57bcfc6969-cpl87     1/1     Running     0          3m43s
saf-default-telemetry-smartgateway-1-deploy   0/1     Completed   0          3m25s
saf-default-telemetry-smartgateway-1-rfczj    1/1     Running     0          3m16s
service-assurance-operator-c8b7f986f-fz6kf    2/2     Running     0          4m25s
smart-gateway-operator-6d7544bf95-mq2xv       2/2     Running     0          5m1s
```

* The expected alertmanager, prometheus, qdr, and SG have appeared and they are NEWER than the SAO pod, proving that they were created by it and not a leftover
* The other operators were installed by OLM in response to the subscription, so may be older than the SAO pod

## Directory agnostic and runs a second time

* I ran it again from inside the `tests/infrared` dir, an illogical choice of pwd

```
[csibbitt@laptop infrared (csibbitt-900-ci-scripts) ]$ pwd
/home/csibbitt/cloudops/service-assurance-operator/tests/infrared
[csibbitt@laptop infrared (csibbitt-900-ci-scripts) ]$ ../../build/build_ci.sh 
Building...
[...]
clusterserviceversion.operators.coreos.com/service-assurance-operator.v0.1.1 created
serviceassurance.infra.watch/saf-default created

[csibbitt@laptop infrared (csibbitt-900-ci-scripts) ]$ echo $?
0

[csibbitt@laptop infrared (csibbitt-900-ci-scripts) ]$ oc get po
NAME                                          READY   STATUS      RESTARTS   AGE
alertmanager-saf-default-0                    2/2     Running     0          54s
interconnect-operator-55b5f8867b-gw5rm        1/1     Running     0          2m1s
prometheus-operator-68f97b8f79-thnkd          1/1     Running     0          2m16s
prometheus-saf-default-0                      3/3     Running     1          58s
saf-default-interconnect-57bcfc6969-pl54p     1/1     Running     0          61s
saf-default-telemetry-smartgateway-1-7flpw    1/1     Running     0          33s
saf-default-telemetry-smartgateway-1-deploy   0/1     Completed   0          42s
service-assurance-operator-c8b7f986f-4jnpz    2/2     Running     0          86s
smart-gateway-operator-6d7544bf95-d7qdn       2/2     Running     0          2m11s
```

## In another namespace

* I tried it in an alternate namespace (from the build directory this time)
```
[csibbitt@laptop build (csibbitt-900-ci-scripts) ]$ ../deploy/remove_saf.sh 
project.project.openshift.io "sa-telemetry" deleted
operatorsource.operators.coreos.com "redhat-service-assurance-operators" deleted
customresourcedefinition.apiextensions.k8s.io "serviceassurances.infra.watch" deleted
customresourcedefinition.apiextensions.k8s.io "smartgateways.smartgateway.infra.watch" deleted
subscription.operators.coreos.com "amq7-cert-manager" deleted
clusterserviceversion.operators.coreos.com "amq7-cert-manager.v1.0.0" deleted
customresourcedefinition.apiextensions.k8s.io "certificates.certmanager.k8s.io" deleted
customresourcedefinition.apiextensions.k8s.io "challenges.certmanager.k8s.io" deleted
customresourcedefinition.apiextensions.k8s.io "clusterissuers.certmanager.k8s.io" deleted
customresourcedefinition.apiextensions.k8s.io "issuers.certmanager.k8s.io" deleted
customresourcedefinition.apiextensions.k8s.io "orders.certmanager.k8s.io" deleted
Waiting for sa-telemetry to disappear
[...]
Waiting for sa-telemetry to disappear
Error from server (NotFound): namespaces "sa-telemetry" not found

[csibbitt@laptop build (csibbitt-900-ci-scripts) ]$ OCP_PROJECT=redhat-stf ./build_ci.sh 
Building...
[...]
serviceassurance.infra.watch/saf-default created

[csibbitt@laptop build (csibbitt-900-ci-scripts) ]$ echo $?
0

[csibbitt@laptop build (csibbitt-900-ci-scripts) ]$ oc project
Using project "redhat-stf" on server "https://api.crc.testing:6443".

[csibbitt@laptop build (csibbitt-900-ci-scripts) ]$ oc get po
NAME                                          READY   STATUS              RESTARTS   AGE
alertmanager-saf-default-0                    2/2     Running             0          25s
interconnect-operator-6b456cc756-wr8wp        1/1     Running             0          115s
prometheus-operator-5766b7658c-jjzgn          1/1     Running             0          108s
prometheus-saf-default-0                      3/3     Running             1          29s
saf-default-interconnect-57bcfc6969-2br45     1/1     Running             0          31s
saf-default-telemetry-smartgateway-1-2d4zh    0/1     ContainerCreating   0          4s
saf-default-telemetry-smartgateway-1-deploy   1/1     Running             0          13s
service-assurance-operator-858bbfcbb9-9w5nc   2/2     Running             0          65s
smart-gateway-operator-6f968965d7-vfv57       2/2     Running             0          119s

[csibbitt@laptop build (csibbitt-900-ci-scripts) ]$ oc describe po service-assurance-operator-858bbfcbb9-9w5nc | grep image
    Image:         image-registry.openshift-image-registry.svc:5000/redhat-stf/service-assurance-operator:latest
    Image ID:      image-registry.openshift-image-registry.svc:5000/redhat-stf/service-assurance-operator@sha256:040b54edd5c8bc3668960455e13a65880ce00edbfe9ff462a4e18b39325679da
```

* And again in the same namespace
```
[csibbitt@laptop build (csibbitt-900-ci-scripts) ]$ OCP_PROJECT=redhat-stf ./build_ci.sh 
Building...
[...]
serviceassurance.infra.watch/saf-default created

[csibbitt@laptop build (csibbitt-900-ci-scripts) ]$ echo $?
0

[csibbitt@laptop build (csibbitt-900-ci-scripts) ]$ oc project
Using project "redhat-stf" on server "https://api.crc.testing:6443".

[csibbitt@laptop build (csibbitt-900-ci-scripts) ]$ oc get po
NAME                                          READY   STATUS              RESTARTS   AGE
interconnect-operator-6b456cc756-lmstp        1/1     Running             0          27s
prometheus-operator-5766b7658c-v62t2          1/1     Running             0          31s
service-assurance-operator-858bbfcbb9-jvplw   0/2     ContainerCreating   0          1s
service-assurance-operator-887db76d9-2hjn4    2/2     Terminating         0          28s
smart-gateway-operator-6f968965d7-kh7mk       0/2     ContainerCreating   0          19s

[csibbitt@laptop build (csibbitt-900-ci-scripts) ]$ oc get po
NAME                                          READY   STATUS              RESTARTS   AGE
alertmanager-saf-default-0                    2/2     Running             0          29s
interconnect-operator-6b456cc756-lmstp        1/1     Running             0          105s
prometheus-operator-5766b7658c-v62t2          1/1     Running             0          109s
prometheus-saf-default-0                      3/3     Running             1          33s
saf-default-interconnect-57bcfc6969-55djl     1/1     Running             0          36s
saf-default-telemetry-smartgateway-1-deploy   1/1     Running             0          17s
saf-default-telemetry-smartgateway-1-zbhbm    0/1     ContainerCreating   0          8s
service-assurance-operator-858bbfcbb9-jvplw   2/2     Running             0          79s
smart-gateway-operator-6f968965d7-kh7mk       2/2     Running             0          97s


```

* In that last test, the first `oc get po` was fast enough for us to see the quickstart SAO being replaced by the locally built SAO before any SAO manifest (or resultant ES/Prom/QDR manifests) are ever installed

TLDR; It works.

**Comment by csibbitt** [comment]
[Re-write local build+deploy+test script for new env](https://trello.com/c/TzgJM7g5/900-re-write-local-builddeploytest-script-for-new-env)

**Comment by csibbitt** [comment]
I added code to wait until everything is Running before we run tests.

## Testing of wait code

```
[csibbitt@laptop service-assurance-operator (csibbitt-900-ci-scripts) ]$ ./build/build_ci.sh | gist

INFO[0000] Building OCI image service-assurance-operator:latest 
INFO[0000] Operator build complete.                     
Error from server (NotFound): namespaces "sa-telemetry" not found
No resources found.
No resources found.
No resources found.
No resources found.
No resources found.
No resources found.
WARNING! Using --password via the CLI is insecure. Use --password-stdin.
WARNING! Your password will be stored unencrypted in /home/csibbitt/.docker/config.json.
Configure a credential helper to remove this warning. See
https://docs.docker.com/engine/reference/commandline/login/#credentials-store

Error from server (NotFound): deployments.apps "saf-default-interconnect" not found
Error from server (NotFound): deployments.apps "saf-default-interconnect" not found
Error from server (NotFound): deploymentconfigs.apps.openshift.io "saf-default-telemetry-smartgateway" not found
https://gist.github.com/f130d5691b173089c2ec2a69e7789f0a
```

This is stderr only, but is actually all okay. Check https://gist.github.com/f130d5691b173089c2ec2a69e7789f0a for the stdout log.

File: build/metadata.sh [hunk 1]
```diff
@@ -0,0 +1,23 @@
+#!/usr/bin/env sh
+#
+# Source this file from other build scripts
+#
+
+# Configure these as necesarry
```
Comment (leifmadsen) [inline build/metadata.sh]: necesarry! sarry.

File: build/update_csv.sh [hunk 2]
```diff
@@ -1,3 +1,4 @@
-#!/usr/bin/env sh
-CSV_VERSION=${CSV_VERSION:-0.1.1}
-operator-sdk olm-catalog gen-csv --csv-version ${CSV_VERSION} --operator-name service-assurance-operator --update-crds
+#!/usr/bin/env bash
+source "$(dirname "$0")/metadata.sh"
+
+operator-sdk olm-catalog gen-csv --csv-version "${CSV_VERSION}" --operator-name service-assurance-operator --update-crds
```
Comment (leifmadsen) [inline build/update_csv.sh]: I would prefer a version check in here because it will fail with !=0.12.0

Comment (csibbitt) [reply build/update_csv.sh]: A version check on operator-sdk? I can work on that.

@pleimer FYI, when working on the build container; we need specifically operator-sdk 0.12.0

**Comment by csibbitt** [comment]
## Testing version check code

```
[csibbitt@laptop build (csibbitt-900-ci-scripts) ]$ operator-sdk version
operator-sdk version: "v0.12.0", commit: "2445fcda834ca4b7cf0d6c38fba6317fb219b469", go version: "go1.13.3 linux/amd64"

[csibbitt@laptop build (csibbitt-900-ci-scripts) ]$ REQUIRED_OPERATOR_SDK_VERSION=v0.133333.0 ./build.sh 

* [info] Checking operator-sdk version

* [error] Wrong operator-sdk version. Wanted v0.133333.0, Got: v0.12.0

[csibbitt@laptop build (csibbitt-900-ci-scripts) ]$ echo $?
1
[csibbitt@laptop build (csibbitt-900-ci-scripts) ]$ ./build.sh 

* [info] Checking operator-sdk version

INFO[0000] Building OCI image service-assurance-operator:latest 
STEP 1: FROM quay.io/operator-framework/ansible-operator:v0.14.1
^C

```

File: build/metadata.sh [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+#
+# Source this file from other build scripts
+#
+
+# Configure these
+OPERATOR_NAME=${OPERATOR_NAME:-service-assurance-operator}
+CSV_VERSION=${CSV_VERSION:-0.1.1}
+IMAGE_BUILDER=${IMAGE_BUILDER:-podman}
+IMAGE_TAG=${IMAGE_TAG:-latest}
+REQUIRED_OPERATOR_SDK_VERSION=${REQUIRED_OPERATOR_SDK_VERSION:-v0.12.0}
```
Comment (leifmadsen) [inline build/metadata.sh]: :1st_place_medal:

File: build/push_container2ocp.sh [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+#
+# Pushes local container image `service-assurance-operator` into OCP
+# * `oc` must already target OCP
+# * container image must already be built. See build.sh for more information
+#
+set -e
+source "$(dirname "$0")/metadata.sh"
+
+if [ "${IMAGE_BUILDER}" = "podman" ]; then
+    REG_EXTRAFLAGS="--tls-verify=false"
```
Comment (leifmadsen) [inline build/push_container2ocp.sh]: :+1:

File: build/push_container2ocp.sh [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+fi
+
+${IMAGE_BUILDER} tag "${OPERATOR_NAME}:${IMAGE_TAG}" "${OCP_REGISTRY}/${OCP_PROJECT}/${OPERATOR_NAME}:${OCP_TAG}"
+${IMAGE_BUILDER} login ${REG_EXTRAFLAGS} -u 'openshift' -p "$(oc whoami -t)" "${OCP_REGISTRY}"
```
Comment (leifmadsen) [inline build/push_container2ocp.sh]: On a non-CRC deployment (e.g. upshift / quicklab) this will fail because the login user is `kubeadmin`.

File: deploy/remove_saf.sh [hunk 6]
```diff
@@ -1,25 +1,30 @@
...
+# Removes SAF and (optionally) amq7 certmanager from your cluster
 #
-SAF_PROJECT=${SAF_PROJECT:-sa-telemetry}
+OCP_PROJECT=${OCP_PROJECT:-sa-telemetry}
+REMOVE_CERTMANAGER=${REMOVE_CERTMANAGER:-true}
 
-# The whole SAF project
-oc delete project "${SAF_PROJECT}"
+# The whole SAF project (start this first since it's slow)
+oc delete project "${OCP_PROJECT}"
```
Comment (leifmadsen) [inline deploy/remove_saf.sh]: I was trying to figure out how to create a `Namespace` object so that we didn't specifically need to rely on the `oc` commands. Never could get it to work though.

File: build/push_container2ocp.sh (hunk 5 above)
Comment (csibbitt) [reply build/push_container2ocp.sh]: Oh yeah? Been awhile since I've had one run to test. Note that the token we are presenting is for the kubeadmin user, it's just that all logins to the registry use the "openshift" user. Not sure why this would be different in CRC but I guess I'm not surprised. I'll try it in quicklab, though.

File: deploy/remove_saf.sh (hunk 6 above)
Comment (csibbitt) [reply deploy/remove_saf.sh]: I played with that and concluded that an openshift project is more than just a namespace and I wasn't going to rip it apart anymore than that.

File: build/push_container2ocp.sh (hunk 5 above)
Comment (leifmadsen) [reply build/push_container2ocp.sh]: I seem to have done it on both my CRC and quicklab tests with the following (although I should probably validate and test in CRC again):

```
REGISTRY=$(oc get route default-route -n openshift-image-registry --template='{{ .spec.host }}')
TOKEN=$(oc whoami -t)
buildah login --tls-verify=false -u kubeadmin -p "${TOKEN}" "${REGISTRY}"
```

Not going to block on this. I can always test post-merge in upshift and file an issue and/or patch.

Comment (csibbitt) [reply build/push_container2ocp.sh]: Ah, neat, kubeadmin does work on CRC! I though that Matthias reported that it didn't.

```
[csibbitt@laptop service-assurance-operator (csibbitt-900-ci-scripts) ]$ ./build/push_container2ocp.sh 
Login Succeeded!
Getting image source signatures
Copying blob 4d253fad236e skipped: already exists
Copying blob ab6008cc8926 skipped: already exists
Copying blob acb86a0a1282 skipped: already exists
Copying blob 3b7c8c69c6ff skipped: already exists
Copying blob c630f5c3e169 skipped: already exists
Copying blob 4bd7408cc1c8 skipped: already exists
Copying blob 911fc1b18694 skipped: already exists
Copying blob 915be7a2cb0f skipped: already exists
Copying blob 1fd5c3a10c1a skipped: already exists
Copying blob 964b072de0ec skipped: already exists
Copying blob 34eed33aab83 skipped: already exists
Copying blob 3fc3a0fdbe1f skipped: already exists
Copying config 3fd22a2569 done
Writing manifest to image destination
Copying config 3fd22a2569 done
Writing manifest to image destination
Storing signatures
```

Still testing with quicklab. I'll break this out into an ENVVAR and set kubeadmin if it works on both.

Comment (csibbitt) [reply build/push_container2ocp.sh]: Quicklab registry is not exposed by default, but I found the instructions for exposing the registry here: https://docs.openshift.com/container-platform/4.3/registry/securing-exposing-registry.html

Curiously, the command they show for logging into the registry doesn't work, but user "openshift" does, as does "kubeadmin". (This was actually what Matthias reported before - we should look into where to log a bug for that)

```
$ echo $KUBECONFIG
/home/csibbitt/.kube/quicklab-stf001

$ podman login -u $(oc whoami) -p $(oc whoami -t) --tls-verify=false $(oc registry info)
Error: error logging into "default-route-openshift-image-registry.apps-crc.testing": invalid username/password

$ podman login -u openshift -p $(oc whoami -t) --tls-verify=false $(oc registry info)
Login Succeeded!

$ podman login -u kubeadmin -p $(oc whoami -t) --tls-verify=false $(oc registry info)
Login Succeeded!

$ oc whoami
kube:admin
```

I'll make the user an envvar anyways, but it doesn't seem like we'll need to change anything. Output of full test coming.

**Comment by csibbitt** [comment]
## Testing on quicklab (Half worked)

* I had to expose the registry first, but it worked just fine with "openshift" user

```
$ echo $KUBECONFIG
/home/csibbitt/.kube/quicklab-stf001

$ oc patch configs.imageregistry.operator.openshift.io/cluster --patch '{"spec":{"defaultRoute":true}}' --type=merge
config.imageregistry.operator.openshift.io/cluster patched

$ QUICKSTART_CONFIG=configs/quicklab.bash ./build/build_ci.sh
[...]
* [info] Pushing new operator image...

Login Succeeded!
[...]
* [info] Waiting for prometheus deployment to complete

Waiting for 1 pods to be ready...
statefulset rolling update complete 1 pods at revision prometheus-saf-default-76d9c98cb5...

* [info] Waiting for alertmanager deployment to complete

Waiting for 1 pods to be ready...
```

This ends up hanging forever because the quicklab.bash has storage fixes for prometheus but not alertmanager. 

The rest of the system is healthy though, so I consider this a success; we just need a quicklab compatible alertmanager manifest.

```
[csibbitt@laptop service-assurance-operator (csibbitt-900-ci-scripts) (⎈ |sa-telemetry/api-stf001-lab-upshift-rdu2-redhat-com:6443/kube:admin)]$ oc get po
NAME                                          READY   STATUS      RESTARTS   AGE
alertmanager-saf-default-0                    0/2     Pending     0          3m46s
interconnect-operator-77fdb4bcc9-g5cxc        1/1     Running     0          5m18s
prometheus-operator-76bb59984d-s6lrd          1/1     Running     0          5m27s
prometheus-saf-default-0                      3/3     Running     1          3m50s
saf-default-interconnect-5dcd4bf7f6-q6ss2     1/1     Running     0          3m52s
saf-default-telemetry-smartgateway-1-deploy   0/1     Completed   0          3m35s
saf-default-telemetry-smartgateway-1-jgpdp    1/1     Running     0          3m26s
service-assurance-operator-fd5548dd-9f5ps     2/2     Running     0          4m18s
smart-gateway-operator-85b6c85b4d-ms9rr       2/2     Running     0          5m25s
[csibbitt@laptop service-assurance-operator (csibbitt-900-ci-scripts) (⎈ |sa-telemetry/api-stf001-lab-upshift-rdu2-redhat-com:6443/kube:admin)]$ oc describe po alertmanager-saf-default-0 | tail
QoS Class:       Burstable
Node-Selectors:  <none>
Tolerations:     node.kubernetes.io/memory-pressure:NoSchedule
                 node.kubernetes.io/not-ready:NoExecute for 300s
                 node.kubernetes.io/unreachable:NoExecute for 300s
Events:
  Type     Reason            Age        From               Message
  ----     ------            ----       ----               -------
  Warning  FailedScheduling  <unknown>  default-scheduler  pod has unbound immediate PersistentVolumeClaims (repeated 4 times)
  Warning  FailedScheduling  <unknown>  default-scheduler  pod has unbound immediate PersistentVolumeClaims (repeated 4 times)
[csibbitt@laptop service-assurance-operator (csibbitt-900-ci-scripts) (⎈ |sa-telemetry/api-stf001-lab-upshift-rdu2-redhat-com:6443/kube:admin)]$ oc get pvc
NAME                                                     STATUS    VOLUME   CAPACITY   ACCESS MODES   STORAGECLASS   AGE
alertmanager-saf-default-db-alertmanager-saf-default-0   Pending                                                     4m
```

File: build/push_container2ocp.sh (hunk 5 above)
Comment (leifmadsen) [reply build/push_container2ocp.sh]: Yea, I had that in my docs as well :)

```
# for use on upshift environments where image registry does not have a default route
# source:  https://docs.openshift.com/container-platform/4.2/registry/securing-exposing-registry.html
 
oc patch configs.imageregistry.operator.openshift.io/cluster --patch '{"spec":{"defaultRoute":true}}' --type=merge
```

Comment (leifmadsen) [reply build/push_container2ocp.sh]: I should have noticed and called that out in my review

**Comment by leifmadsen** [comment]
So I'm running the script right now, and as my first attempt, I'm not exactly sure how long this is expected to run, and if it'll just wait indefinitely.

I think this is an RFE, but I feel like there should be some sort of timeout setting where a bunch of data is dumped at the end, or some sort of periodic `oc get csv` perhaps? The CSV basically will rotate around its status to Installing, Succeeded, etc.

On another screen I've tried running that and it doesn't look like things are progressing. At this point I Just have `Waiting for SAO...` for what I'm going to assume is indefinitely. I'll continue looking at this tomorrow, just wanted to throw up a hail mary attempt before heading off for the night.

**Comment by leifmadsen** [comment]
Had to wait on a buddy for a game, so I had a chance to run the `remove_saf.sh` and `QUICKSTART_CONFIG=configs/quicklab.bash ./quickstart.sh` and things are moving along, and the script has completed successfully.

Everything looks good. I have an adjustment to make to the quickstart.bash but I'll do that in another branch.

**Comment by csibbitt** [comment]
> So I'm running **the script** right now, and as my first attempt, I'm not exactly sure how long this is expected to run, and if it'll just wait indefinitely.... At this point I Just have `Waiting for SAO...` for what I'm going to assume is indefinitely. I

Points of order - Which script did you run? What OCP did you run it against? What output led up to the failure? I really appreciate the feedback, but I would appreciate it more if it came with data I can use to help.

> I think this is an RFE, but I feel like there should be some sort of timeout setting where a bunch of data is dumped at the end, or some sort of periodic `oc get csv` perhaps? The CSV basically will rotate around its status to Installing, Succeeded, etc.

Yes, this appears to be an RFE for quickstart; which is not really what I'm working on. I like the idea, for sure. Do you have an idea of what we should output? `oc get csv` isn't too useful. A describe? We're going to end up with thousands of extra lines of ourput in our CI jobs as a result, so we should make sure we pick something useful. If you could re-run your tests and add the output that is useful for you to debug the problem you're running into, that would certainly be welcome.

> remove_saf.sh and QUICKSTART_CONFIG=configs/quicklab.bash ./quickstart.sh

Is that what you ran the first time when it hung forever, or were you running build_ci.sh?

**Comment by leifmadsen** [comment]
> > So I'm running **the script** right now, and as my first attempt, I'm not exactly sure how long this is expected to run, and if it'll just wait indefinitely.... At this point I Just have `Waiting for SAO...` for what I'm going to assume is indefinitely. I
> 
> Points of order - Which script did you run? What OCP did you run it against? What output led up to the failure? I really appreciate the feedback, but I would appreciate it more if it came with data I can use to help.

Fair points! I guess I was just trying to note something quickly after hours, and it appears to not be useful :)

"The script" == `quickstart.sh` in the `deploy/` directory. I ran it against a 1x1 QuickLAB OCP 4.2.16 instance. The script was originally run as `./deploy/quickstart.sh` as I saw that it _should_ work from that position (and I think it did?).

There wasn't a whole lot of data to provide. It looks like it executed and moved down to the `Waiting for SAO...` but I only ever saw the AMQ Cert Manager subscription, and the OperatorSource for redhat-service-assurance-operators. I did not see an OperatorGroup, and I didn't see the PackageManifests loaded for redhat-service-assurance-operators.

Since it worked the next try, I'm going to chalk it up to environmental issues and not the script itself unless I can reproduce it on a fresh OCP4 instance.

> 
> > I think this is an RFE, but I feel like there should be some sort of timeout setting where a bunch of data is dumped at the end, or some sort of periodic `oc get csv` perhaps? The CSV basically will rotate around its status to Installing, Succeeded, etc.
> 
> Yes, this appears to be an RFE for quickstart; which is not really what I'm working on. I like the idea, for sure. Do you have an idea of what we should output? `oc get csv` isn't too useful. A describe? We're going to end up with thousands of extra lines of ourput in our CI jobs as a result, so we should make sure we pick something useful. If you could re-run your tests and add the output that is useful for you to debug the problem you're running into, that would certainly be welcome.
> 
> > remove_saf.sh and QUICKSTART_CONFIG=configs/quicklab.bash ./quickstart.sh
> 
> Is that what you ran the first time when it hung forever, or were you running build_ci.sh?

I ran the `quickstart.sh` not the `build_ci.sh`, which now that I've looked again, while there are changed in `quickstart.sh` you are indeed correct, that's not the area you're working on :)

Sorry for the noise, this doesn't appear to be the correct PR for comments on quickstart deploys.

**Comment by csibbitt** [comment]
> 1x1 QuickLAB

Can't point the finger definitively at this point, but I stopped using single node quicklabs because I noticed pods getting OOMKilled. I guess CRC comes with less stuff preconfigured, so I don't run into trouble with a 16G CRC, but the 16G quicklab nodes have definitely given me trouble frequently.

**Comment by leifmadsen** [comment]
> > 1x1 QuickLAB
> 
> Can't point the finger definitively at this point, but I stopped using single node quicklabs because I noticed pods getting OOMKilled. I guess CRC comes with less stuff preconfigured, so I don't run into trouble with a 16G CRC, but the 16G quicklab nodes have definitely given me trouble frequently.

Yea it's been hit and miss. I've been deploying 1 master + 1 compute nodes recently, and I've not been having issues with full events + metrics deployments. I have seen that OOMKilled thing though, and it almost feels like it's an underlying infrastructure memory issue.

I had been going back to 1x3 deployments, but you can't even run 2 of those in the current quota due to needing a bootstrap (undercloud equiv) node for the deployments, which counts towards your quota. That means your 4 node deployment is actually a 5 node deployment. I ran out of quota from even running a single 1x1 cluster next to my old 1x3 deployment.

In this case, it wasn't an OOMKilled issue, because it didn't even get far enough to having the objects loaded into memory. I'm going to pretend it didn't happen until I see it happen again.

**Comment by leifmadsen** [comment]
# Testing

Environment:

* podman
  * Version:            1.7.0
* operator-sdk
  * operator-sdk version: "v0.12.0", commit: "2445fcda834ca4b7cf0d6c38fba6317fb219b469", go version: "go1.13.3 linux/amd64"

## `build.sh`

```
./build.sh 

* [info] Checking operator-sdk version

INFO[0000] Building OCI image service-assurance-operator:latest 
STEP 1: FROM quay.io/operator-framework/ansible-operator:v0.14.1
Getting image source signatures
Copying blob fd8daf2668d1 done  
Copying blob b0a6031c09c8 done  
Copying blob cb3c77f9bdd8 done  
Copying blob 2b6a2ff03bbe done  
Copying blob 725bd73625cc done  
Copying blob 35105e67a8c9 done  
Copying blob a4c437a72be2 done  
Copying blob 9b8b2a1369c4 done  
Copying blob 2bc41d47dd50 done  
Copying blob fb8dcb3732a4 done  
Copying config c84a46d219 done  
Writing manifest to image destination
Storing signatures
STEP 2: COPY watches.yaml ${HOME}/watches.yaml
785b8f32e56fd4c2a69423c8dc35ee2ecf00843a440c4e6bb4df53ff69cb2037
STEP 3: COPY roles/ ${HOME}/roles/
STEP 4: COMMIT service-assurance-operator:latest
fd35ba1313e9c6f52ed75eece3c062a3154e1331608b4404dd1a7fedff458668
fd35ba1313e9c6f52ed75eece3c062a3154e1331608b4404dd1a7fedff458668
INFO[0022] Operator build complete.                     

podman images
REPOSITORY                                    TAG       IMAGE ID       CREATED         SIZE
localhost/service-assurance-operator          latest    fd35ba1313e9   6 seconds ago   505 MB
quay.io/operator-framework/ansible-operator   v0.14.1   c84a46d21985   9 days ago      505 MB

## `build_ci.sh`

Looks good so far. I'm testing it against a quicklab host, and that doesn't work since it's using the built in manifests (as I would expect for a valid CI test). Since those are not valid for a quicklab environment things just sit pending indefinitely.

_At this point I will drop testing against quicklab since that isn't in scope for this PR_

**Comment by csibbitt** [comment]
I'm going to merge these now. No one else is commenting, and if they are broken (don't seem to be!) we can always fix them.
